src/lang/std/artifactGraph.test.ts

```diff
@@ -0,0 +1,743 @@
+          return [`UUID-${index}`, stableValue]
+        })
+      )
+      expect(stableMap).toMatchSnapshot()
+    })
+
+    it('screenshot graph', async () => {
+      // Ostensibly this takes a screen shot of the graph of the artifactGraph
+      // but it's it also tests that all of the id links are correct because if one
+      // of the edges refers to a non-existent node, the graph will throw.
+      // further more we can check that each edge is bi-directional, if it's not
+      // by checking the arrow heads going both ways, on the graph.
+      await GraphTheGraph(theMap, 1400, 1400, 'exampleCode1.png')
+    }, 20000)
+  })
+})
+
+describe('capture graph of sketchOnFaceOnFace...', () => {
+  describe('code with an extrusion, fillet and sketch of face:', () => {
+    let ast: Program
+    let theMap: ReturnType<typeof createArtifactGraph>
+    it('setup', async () => {
+      // putting this logic in here because describe blocks runs before beforeAll has finished
+      const {
+        orderedCommands,
+        responseMap,
+        ast: _ast,
+      } = getCommands('sketchOnFaceOnFaceEtc')
+      ast = _ast
+      theMap = createArtifactGraph({ orderedCommands, responseMap, ast })
+
+      // Ostensibly this takes a screen shot of the graph of the artifactGraph
+      // but it's it also tests that all of the id links are correct because if one
+      // of the edges refers to a non-existent node, the graph will throw.
+      // further more we can check that each edge is bi-directional, if it's not
+      // by checking the arrow heads going both ways, on the graph.
+      await GraphTheGraph(theMap, 2500, 2500, 'sketchOnFaceOnFaceEtc.png')
+    }, 20000)
+  })
+})
+
+function getCommands(codeKey: CodeKey): CacheShape[CodeKey] & { ast: Program } {
+  const ast = parse(codeKey)
+  if (err(ast)) {
+    console.error(ast)
+    throw ast
+  }
+  const file = fs.readFileSync(fullPath, 'utf-8')
+  const parsed: CacheShape = JSON.parse(file)
+  // these either already exist from the last run, or were created in
+  const orderedCommands = parsed[codeKey].orderedCommands
+  const responseMap = parsed[codeKey].responseMap
+  return {
+    orderedCommands,
+    responseMap,
+    ast,
+  }
+}
+
+async function GraphTheGraph(
+  theMap: ArtifactGraph,
+  sizeX: number,
+  sizeY: number,
+  imageName: string
+) {
+  const nodes: Array<{ id: string; label: string }> = []
+  const edges: Array<{ source: string; target: string; label: string }> = []
+  let index = 0
+  for (const [commandId, artifact] of theMap) {
+    nodes.push({
+      id: commandId,
+      label: `${artifact.type}-${index++}`,
+    })
+    Object.entries(artifact).forEach(([propName, value]) => {
+      if (
+        propName === 'type' ||
+        propName === 'codeRef' ||
+        propName === 'subType'
+      )
+        return
+      if (Array.isArray(value))
+        value.forEach((v) => {
+          edges.push({ source: commandId, target: v, label: propName })
```

Comment:
Should this check that each item is a string and truthy before inserting?

Reply:
Think it's probably fine because there shouldn't be empty strings in the array, also this code will error if you add an edge to a node that doesn't exist, so it would need to pair with a node without a command id.

But better to be safe than sorry so I did put in a check.

Reply:
In my opinion, it's not about "better safe than sorry"; it's about allowing it to type check with real types. Due to legacy JS, the return type of `Array.isArray()` is incorrectly `any[]`. It should be `unknown[]`. Open issues linked below. I made a helper in #3207 that I hope to merge and convert all our code over at some point.

https://github.com/microsoft/TypeScript/issues/17002
https://github.com/microsoft/TypeScript/issues/33700